README
NAME
    Perl::Signature - Generate functional signatures for Perl source code

  DESCRIPTION
    In early beta, PPI introduced the concept of "Document Normalization"
    into the core. It had previously only been implemented "behind the
    scenes" as part of Perl::Compare.

    Unfortunately, there isn't a whole lot of things you can do with a
    PPI::Document::Normalized object. It's a giant twisty mass of objects
    and perl structure, and not very practical for long term storage.

    Perl::Signature implements the idea of a "functional signature" for Perl
    documents, implemented in a similar way to Object::Signature.

    The normalized document is serialized to a string with Storable, then
    this string is converted into a MD5 hash, producing a single short
    string which represents the functionality of the Perl document.

    This signature can then be stored and transfered easily, and at any
    later point the signature can be regenerated for the file to ensure that
    it has not changed (functionally).

  Not Stable Across Upgrades
    Perl::Signature is relatively sensitive to change.

    Primarily, this is because PPI::Normal is biased towards false negative
    comparison. (Avoiding false "these are the same" by accepting a number
    of false "these are not the same" results).

    In addition, the serialization of Storable is not assured to use
    identical file formats across versions.

    In short, you should assume that a signature is valid at best for only
    as long as the PPI and Storable versions are the same, and at worst only
    for the current process.

METHODS
    PPI::Signature provides two sets of methods. A set of

  file_signature $filename
    The "file_signature" static method takes a filename and produces a
    signature for the file.

    Returns a 32 character hexidecimal MD5 signature, or "undef" on error.

  source_signature $content | \$content
    The "source_signature" static method generates a signature for any
    arbitrary Perl source code, which can be passed as either a raw string,
    or a reference to a SCALAR containing the code.

    Returns a 32 character hexidecimal MD5 signature, or "undef" on error.

  document_signature $Document
    The "document_signature" method takes a PPI::Document object and
    generates a signature for it.

    Returns a 32 character hexidecimal MD5 signature, or "undef" on error.

  new $file
    As well as static methods for generatic signatures, Perl::Signature also
    provides a simple way to create signature objects for a particular file.

    This makes it relatively easy to see if a file has changed

    The "new" constructor takes as argument the name of a file, and creates
    an object that remembers current signature of the file.

  file
    The "file" accessor returns the name of the file that a Perl::Signature
    object is set to.

  current
    The "current" method returns the current signature for the file.

    Returns a 32 character hexidecimal MD5 signature, or "undef" on error.

  original
    The "original" accessor returns the original signature at the time of
    the creation of the object.

  changed
    The "changed" method checks to see if the signature has changed since
    the object was created.

    Returns true if the file has been (functionally) changed, false if not,
    or "undef" on error.

  unchanged
    The "unchanged" method checks to ensure that the signature has not
    changed since the object was created.

    Returns true if the file is (functionally) unchanged, false if it has
    changed, or "undef" on error.

SUPPORT
    All bugs should be filed via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Perl-Signature>

    For other issues, or commercial enhancement or support, contact the
    author.

AUTHORS
    Adam Kennedy <cpan@ali.as>

SEE ALSO
    <http://ali.as/>, PPI

COPYRIGHT
    Copyright (c) 2004 - 2006 Adam Kennedy. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension Perl::Signature

0.07 Fri Dec 30 2005
	- ::Set->new($layer) only accepted 1. Now accepts 2 as well. :/
	- Updated PPI dep to 1.109 to fix $_ being clobbered.
	- Upgraded Module::Install

0.05 Wed May 25 2005
	- Updated to handle API change at PPI 0.990

0.04  Wed Feb  9 2005
	- Added the ->files and ->file methods

0.03  Tue Feb  8 2005
	- Added Perl::Signature::Set

0.02  Mon Feb  7 2005
	- Upgraded to the beta PPI code
	- API changed quite a bit
	- Added instantiable usage

0.01  Mon Sep 22 2004
	- Created the initial implementation


inc/Module/Install.pm
#line 1 "/home/adam/parseperl/Perl-Signature/inc/Module/Install.pm - /usr/local/share/perl/5.8.4/Module/Install.pm"
package Module::Install;
use 5.004;

$VERSION = '0.50'; # Don't forget to update Module::Install::Admin too!

die << "." unless $INC{join('/', inc => split(/::/, __PACKAGE__)).'.pm'};
Please invoke ${\__PACKAGE__} with:

    use inc::${\__PACKAGE__};

not:

    use ${\__PACKAGE__};

.

use strict 'vars';
use Cwd qw(cwd abs_path);
use FindBin;
use File::Find ();
use File::Path ();

@inc::Module::Install::ISA = 'Module::Install';
*inc::Module::Install::VERSION = *VERSION;

sub autoload {
    my $self   = shift;
    my $caller = $self->_caller;

    my $cwd = cwd();
    my $sym = "$caller\::AUTOLOAD";

    $sym->{$cwd} = sub {
        my $pwd = cwd();
        if (my $code = $sym->{$pwd}) {
            goto &$code unless $cwd eq $pwd; # delegate back to parent dirs
        }
        $$sym =~ /([^:]+)$/ or die "Cannot autoload $caller - $sym";
        unshift @_, ($self, $1);
        goto &{$self->can('call')} unless uc($1) eq $1;
    };
}

sub import {
    my $class = shift;
    my $self = $class->new(@_);

    if (not -f $self->{file}) {
        require "$self->{path}/$self->{dispatch}.pm";
        File::Path::mkpath("$self->{prefix}/$self->{author}");
        $self->{admin} = 
          "$self->{name}::$self->{dispatch}"->new(_top => $self);
        $self->{admin}->init;
        @_ = ($class, _self => $self);
        goto &{"$self->{name}::import"};
    }

    *{$self->_caller . "::AUTOLOAD"} = $self->autoload;
    $self->preload;

    # Unregister loader and worker packages so subdirs can use them again
    delete $INC{"$self->{file}"};
    delete $INC{"$self->{path}.pm"};
}

sub preload {
    my ($self) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    my @exts = @{$self->{extensions}};

    unless (@exts) {
        my $admin = $self->{admin};
        @exts = $admin->load_all_extensions;
    }

    my %seen_method;
    foreach my $obj (@exts) {
        while (my ($method, $glob) = each %{ref($obj) . '::'}) {
            next unless defined *{$glob}{CODE};
            next if $method =~ /^_/;
            next if $method eq uc($method);
            $seen_method{$method}++;
        }
    }

    my $caller = $self->_caller;
    foreach my $name (sort keys %seen_method) {
        *{"${caller}::$name"} = sub {
            ${"${caller}::AUTOLOAD"} = "${caller}::$name";
            goto &{"${caller}::AUTOLOAD"};
        };
    }
}

sub new {
    my ($class, %args) = @_;

    # ignore the prefix on extension modules built from top level.
    my $base_path = abs_path($FindBin::Bin);
    delete $args{prefix} unless abs_path(cwd()) eq $base_path;

    return $args{_self} if $args{_self};

    $args{dispatch} ||= 'Admin';
    $args{prefix}   ||= 'inc';
    $args{author}   ||= '.author';
    $args{bundle}   ||= 'inc/BUNDLES';
    $args{base}     ||= $base_path;

    $class =~ s/^\Q$args{prefix}\E:://;
    $args{name}     ||= $class;
    $args{version}  ||= $class->VERSION;

    unless ($args{path}) {
        $args{path}  = $args{name};
        $args{path}  =~ s!::!/!g;
    }
    $args{file}     ||= "$args{base}/$args{prefix}/$args{path}.pm";

    bless(\%args, $class);
}

sub call {
    my $self   = shift;
    my $method = shift;
    my $obj    = $self->load($method) or return;

    unshift @_, $obj;
    goto &{$obj->can($method)};
}

sub load {
    my ($self, $method) = @_;

    $self->load_extensions(
        "$self->{prefix}/$self->{path}", $self
    ) unless $self->{extensions};

    foreach my $obj (@{$self->{extensions}}) {
        return $obj if $obj->can($method);
    }

    my $admin = $self->{admin} or die << "END";
The '$method' method does not exist in the '$self->{prefix}' path!
Please remove the '$self->{prefix}' directory and run $0 again to load it.
END

    my $obj = $admin->load($method, 1);
    push @{$self->{extensions}}, $obj;

    $obj;
}

sub load_extensions {
    my ($self, $path, $top_obj) = @_;

    unshift @INC, $self->{prefix}
        unless grep { $_ eq $self->{prefix} } @INC;

    local @INC = ($path, @INC);
    foreach my $rv ($self->find_extensions($path)) {
        my ($file, $pkg) = @{$rv};
        next if $self->{pathnames}{$pkg};

        local $@;
        my $new = eval { require $file; $pkg->can('new') };
        if (!$new) { warn $@ if $@; next; }
        $self->{pathnames}{$pkg} = delete $INC{$file};
        push @{$self->{extensions}}, &{$new}($pkg, _top => $top_obj );
    }

    $self->{extensions} ||= [];
}

sub find_extensions {
    my ($self, $path) = @_;
    my @found;

    File::Find::find(sub {
        my $file = $File::Find::name;
        return unless $file =~ m!^\Q$path\E/(.+)\.pm\Z!is;
        return if $1 eq $self->{dispatch};

        $file = "$self->{path}/$1.pm";
        my $pkg = "$self->{name}::$1"; $pkg =~ s!/!::!g;
        push @found, [$file, $pkg];
    }, $path) if -d $path;

    @found;
}

sub _caller {
    my $depth = 0;
    my $caller = caller($depth);

    while ($caller eq __PACKAGE__) {
        $depth++;
        $caller = caller($depth);
    }

    $caller;
}

1;


lib/Perl/Signature/Set.pm
package Perl::Signature::Set;

=pod

=head1 NAME

Perl::Signature::Set - Create, store and check groups of signatures

=head1 DESCRIPTION

There are a number of cases where you might want to create and look after
a whole bunch of signatures.

The most common of these is:

  1. Generate signatures
  2. Do some process that shouldn't change the files functionally
  3. Test to make sure it didn't

Examples for 2. could be things like applying L<Perl::Tidy>, merging in
documentation-only patches from external sources, and other similar
things.

Perl::Signature::Set lets you create an object that can store a while
bunch of file signatures, save the set to a file, load it in again,
and test the lot to check for changes.

=head2 Saving and Loading

For simplicity and easy of creation, Perl::Signature::Set has been
implemented as a subclass of L<Config::Tiny>.

=head1 METHODS

=cut

use strict;
use base 'Config::Tiny';
use Perl::Signature ();

use vars qw{$VERSION $errstr};
BEGIN {
	$VERSION = '0.07';
	$errstr  = '';
}





#####################################################################
# Constructor and Accessors

=pod

=head2 new

Creates a new Perl::Signature::Set object. Takes as an optional argument
the normalization layer you wish to use.

Returns a new Perl::Signature::Set object.

=cut

sub new {
	my $class = ref $_[0] ? ref shift : shift;
	my $layer = @_ ? (defined $_[0] and $_[0] =~ /^[12]$/) ? shift : return undef : 1;

	# Create the basic object
	my $self = bless {
		signature => {
			layer => $layer,
			},
		files => {},
		}, $class;

	$self;
}

=pod

=head2 layer

The C<layer> accessor returns the normalization layer that was used for
all of the signatures in the object.

=cut

sub layer { $_[0]->{signature}->{layer} }





#####################################################################
# Perl::Signature Methods

=pod

=head2 add $file

The C<add> method takes the name of a file to generate a signature for
and add to the set.

Returns the actual L<Perl::Signature> object created as a convenience,
or C<undef> if the file has already been added, or on error.

=cut

sub add {
	my $self = shift;
	my $file = -f $_[0] ? shift : return undef;
	return undef if $self->{files}->{$file};

	# Create the Signature object, and add it
	my $Signature = Perl::Signature->new( $file ) or return undef;
	$self->{files}->{$file} = $Signature;
}

=pod

=head2 files

The C<files> method provides all of the names of the files contained
in the set, in default sorted order.

Returns a list of file names, or the null list if the set contains no
files.

=cut

sub files {
	my $self  = shift;
	my @files = sort keys %{$self->{files}};
	@files;
}

=pod

=head2 file $filename

The C<file> method is used to get the L<Perl::Signature> object for a
single named file.

Returns a L<Perl::Signature> object, or C<undef> if the file is not in
the set.

=cut

sub file {
	my $self = shift;
	my $file = defined $_[0] ? shift : return undef;
	$self->{files}->{$file};
}

=pod

=head2 signatures

The C<signatures> method returns all of the Signature objects from
the Set, in filename-sorted order.

Returns a list of L<Perl::Signature> objects, or the null list if
the set does not contain any Signature objects.

=cut

sub signatures {
	my $self  = shift;
	my $files = $self->{files};
	map { $files->{$_} } sort keys %$files;
}

=pod

=head2 changes

The C<changes> method checks the signatures for each file and provides
a hash listing the files that have changed as the key,
and either "changed" or "removed" as the value.

Returns a HASH reference, false (C<''>) if there are no changes, or
C<undef> on error.

=cut

sub changes {
	my $self = shift;

	# Iterate of the files and check each one
	my %results = ();
	foreach my $file ( keys %{$self->{files}} ) {
		if ( -f $file ) {
			my $changed = $self->{files}->{$file}->changed;
			return undef unless defined $changed;
			$results{$file} = 'changed' if $changed;
		} else {
			$results{$file} = 'removed';
		}
	}

	keys %results ? \%results : '';
}





#####################################################################
# Config::Tiny Methods

sub read_string {
	my $class = shift;

	# Create the basic object using the parent method
	my $self = $class->SUPER::read_string(@_);

	# Check and clean up
	$self->{signature}          or return undef;
	$self->{signature}->{layer} or return undef;
	$self->{files}              or return undef;

	# Manually bless a signature object for each file entry
	my $files = $self->{files};
	foreach my $file ( keys %$files ) {
		my $signature = $files->{$file};
		$signature =~ /^[a-f0-9]{32}$/ or return undef;
		$files->{$file} = bless {
			file      => $file,
			signature => $signature,
			}, 'Perl::Signature';
	}

	$self;
}

sub write_string {
	my $self = shift;

	# Create the equivalent Config::Tiny object
	my $save = Config::Tiny->new;
	$save->{signature}->{layer} = $self->{signature}->{layer};
	foreach my $file ( keys %{$self->{files}} ) {
		$save->{files}->{$file} = $self->{files}->{$file}->original;
	}

	$save->write_string;
}

sub errstr { $errstr }
sub _error { $errstr = $_[1]; undef }

1;

=pod

=head1 SUPPORT

All bugs should be filed via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Perl-Signature>

For other issues, or commercial enhancement or support, contact the author.

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 SEE ALSO

L<PPI>, L<Perl::Signature>, L<Perl::Compare>

=head1 COPYRIGHT

Copyright (c) 2005 Adam Kennedy. All rights reserved.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


lib/Perl/Signature.pm
package Perl::Signature;

=pod

=head1 NAME

Perl::Signature - Generate functional signatures for Perl source code

=head2 DESCRIPTION

In early beta, L<PPI> introduced the concept of "Document Normalization"
into the core. It had previously only been implemented "behind the scenes"
as part of L<Perl::Compare>.

Unfortunately, there isn't a whole lot of things you can do with a
L<PPI::Document::Normalized> object. It's a giant twisty mass of objects
and perl structure, and not very practical for long term storage.

L<Perl::Signature> implements the idea of a "functional signature" for
Perl documents, implemented in a similar way to L<Object::Signature>.

The normalized document is serialized to a string with L<Storable>, then
this string is converted into a MD5 hash, producing a single short string
which represents the functionality of the Perl document.

This signature can then be stored and transfered easily, and at any later
point the signature can be regenerated for the file to ensure that it has
not changed (functionally).

=head2 Not Stable Across Upgrades

Perl::Signature is relatively sensitive to change.

Primarily, this is because L<PPI::Normal> is biased towards false
negative comparison. (Avoiding false "these are the same" by accepting a
number of false "these are not the same" results).

In addition, the serialization of L<Storable> is not assured to use
identical file formats across versions.

In short, you should assume that a signature is valid at best for only
as long as the PPI and Storable versions are the same, and at worst only
for the current process.

=head1 METHODS

PPI::Signature provides two sets of methods. A set of 

=cut

use 5.005;
use strict;
use PPI           ();
use PPI::Util     '_Document';
use Storable      ();
use Digest::MD5   ();

use vars qw{$VERSION};
BEGIN {
	$VERSION = '0.07';
}





#####################################################################
# Static Methods

=pod

=head2 file_signature $filename

The C<file_signature> static method takes a filename and produces a
signature for the file.

Returns a 32 character hexidecimal MD5 signature, or C<undef> on error.

=cut

sub file_signature {
	my $class    = ref $_[0] ? ref shift : shift;
	my $filename = -f $_[0]  ? shift : return undef;
	my $Document = PPI::Document->new( $filename ) or return undef;
	$class->document_signature( $Document );
}

=pod

=head2 source_signature $content | \$content

The C<source_signature> static method generates a signature for any
arbitrary Perl source code, which can be passed as either a raw string,
or a reference to a SCALAR containing the code.

Returns a 32 character hexidecimal MD5 signature, or C<undef> on error.

=cut

sub source_signature {
	my $class  = ref $_[0] ? ref shift : shift;
	my $source = defined $_[0] ? shift : return undef;
	$source    = $$source if ref $source;

	# Build the PPI::Document
	my $Document = PPI::Document->new( \$source ) or return undef;
	$class->document_signature( $Document );
}

=pod

=head2 document_signature $Document

The C<document_signature> method takes a L<PPI::Document> object and
generates a signature for it.

Returns a 32 character hexidecimal MD5 signature, or C<undef> on error.

=cut

sub document_signature {
	my $class    = ref $_[0] ? ref shift : shift;
	my $Document = _Document(shift) or return undef;

	# Normalize the PPI::Document
	my $Normalized = $Document->normalized or return undef;

	# Freeze the normalized document
	my $string = Storable::freeze $Normalized;
	return undef unless defined $string;

	# Last step, hash the string
	Digest::MD5::md5_hex( $string ) or undef;
}





#####################################################################
# Object Methods

=pod

=head2 new $file

As well as static methods for generatic signatures, L<Perl::Signature>
also provides a simple way to create signature objects for a particular
file.

This makes it relatively easy to see if a file has changed

The C<new> constructor takes as argument the name of a file, and creates
an object that remembers current signature of the file.

=cut

sub new {
	my $class = ref $_[0] ? ref shift : shift;
	my $file  = -f $_[0] ? shift : return undef;

	# Get the current signature for the file
	my $signature = $class->file_signature( $file ) or return undef;

	# Create the object
	my $self = bless {
		file      => $file,
		signature => $signature,
		}, $class;

	$self;
}

=pod

=head2 file

The C<file> accessor returns the name of the file that a Perl::Signature
object is set to.

=cut

sub file { $_[0]->{file} }

=pod

=head2 current

The C<current> method returns the current signature for the file.

Returns a 32 character hexidecimal MD5 signature, or C<undef> on error.

=cut

sub current {
	my $self = shift;
	-f $self->file or return undef;
	$self->file_signature( $self->file );
}

=pod

=head2 original

The C<original> accessor returns the original signature at the time of
the creation of the object.

=cut

sub original { $_[0]->{signature} }

=pod

=head2 changed

The C<changed> method checks to see if the signature has changed since
the object was created.

Returns true if the file has been (functionally) changed, false if not,
or C<undef> on error.

=cut

sub changed {
	my $self    = shift;
	my $current = $self->current or return undef;
	$current ne $self->original;	
}

=pod

=head2 unchanged

The C<unchanged> method checks to ensure that the signature has not
changed since the object was created.

Returns true if the file is (functionally) unchanged, false if it has
changed, or C<undef> on error.

=cut

sub unchanged {
	my $self    = shift;
	my $current = $self->current or return undef;
	$current eq $self->original;
}

1;

=pod

=head1 TO DO

- Write Perl::Signature::Set, to handle multiple signatures

=head1 SUPPORT

All bugs should be filed via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Perl-Signature>

For other issues, or commercial enhancement or support, contact the author.

=head1 AUTHORS

Adam Kennedy (Maintainer), L<http://ali.as/>, cpan@ali.as

=head1 COPYRIGHT

Copyright (c) 2004 - 2005 Adam Kennedy. All rights reserved.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
